File: aitrailblazer.AppHost/go-sec-edgar-ws/company_info.go
package main

import (
	"encoding/json"
	"errors"
	"io/ioutil"
	"strings"
)

type CompanyInfo struct {
	Fields  []string        `json:"fields"`
	Records [][]interface{} `json:"data"`
}

func NewCompanyInfo(jsonPath string) (*CompanyInfo, error) {
	data, err := ioutil.ReadFile(jsonPath)
	if err != nil {
		return nil, err
	}

	var companyInfo CompanyInfo
	err = json.Unmarshal(data, &companyInfo)
	if err != nil {
		return nil, err
	}

	return &companyInfo, nil
}

func (ci *CompanyInfo) ToDataFrame() map[string][]interface{} {
	df := make(map[string][]interface{})
	for _, field := range ci.Fields {
		df[field] = make([]interface{}, len(ci.Records))
	}

	for i, record := range ci.Records {
		for j, field := range ci.Fields {
			df[field][i] = record[j]
		}
	}

	return df
}

func (ci *CompanyInfo) GetCIKByTicker(ticker string) (string, error) {
	df := ci.ToDataFrame()
	for i, t := range df["ticker"] {
		if t == ticker {
			return df["cik"][i].(string), nil
		}
	}
	return "", errors.New("no company found with ticker: " + ticker)
}

func (ci *CompanyInfo) GetNameByTicker(ticker string) (string, error) {
	df := ci.ToDataFrame()
	for i, t := range df["ticker"] {
		if t == ticker {
			return df["name"][i].(string), nil
		}
	}
	return "", errors.New("no company found with ticker: " + ticker)
}

func (ci *CompanyInfo) GetExchangeByTicker(ticker string) (string, error) {
	df := ci.ToDataFrame()
	for i, t := range df["ticker"] {
		if t == ticker {
			return df["exchange"][i].(string), nil
		}
	}
	return "", errors.New("no company found with ticker: " + ticker)
}

func (ci *CompanyInfo) GetTickersByExchange(exchange string) []string {
	df := ci.ToDataFrame()
	var tickers []string
	for i, e := range df["exchange"] {
		if strings.EqualFold(e.(string), exchange) {
			tickers = append(tickers, df["ticker"][i].(string))
		}
	}
	return tickers
}

func (ci *CompanyInfo) GetAllExchanges() []string {
	df := ci.ToDataFrame()
	exchangeMap := make(map[string]bool)
	for _, e := range df["exchange"] {
		exchangeMap[e.(string)] = true
	}

	var exchanges []string
	for exchange := range exchangeMap {
		exchanges = append(exchanges, exchange)
	}
	return exchanges
}

func (ci *CompanyInfo) GetCompaniesByExchange(exchange string) ([]map[string]interface{}, error) {
	df := ci.ToDataFrame()
	var companies []map[string]interface{}
	for i, e := range df["exchange"] {
		if e == exchange {
			company := map[string]interface{}{
				"cik":    df["cik"][i],
				"name":   df["name"][i],
				"ticker": df["ticker"][i],
			}
			companies = append(companies, company)
		}
	}
	if len(companies) == 0 {
		return nil, errors.New("no companies found for exchange: " + exchange)
	}
	return companies, nil
}

func (ci *CompanyInfo) SearchByName(substring string) []map[string]interface{} {
	df := ci.ToDataFrame()
	var results []map[string]interface{}
	for i, name := range df["name"] {
		if strings.Contains(strings.ToLower(name.(string)), strings.ToLower(substring)) {
			result := map[string]interface{}{
				"cik":    df["cik"][i],
				"name":   df["name"][i],
				"ticker": df["ticker"][i],
			}
			results = append(results, result)
		}
	}
	return results
}

File: aitrailblazer.AppHost/sec-edgar-ws/sec_data/company_info.py
# company_info.py
import pandas as pd
import json

class CompanyInfo:
    def __init__(self, json_path):
        with open(json_path, "r") as file:
            self.data = json.load(file)

        self.fields = self.data["fields"]
        self.records = self.data["data"]

    def to_dataframe(self):
        return pd.DataFrame(self.records, columns=self.fields)

    def get_cik_by_ticker(self, ticker):
        df = self.to_dataframe()
        result = df[df["ticker"] == ticker]
        if not result.empty:
            return result.iloc[0]["cik"]
        else:
            raise ValueError(f"No company found with ticker: {ticker}")
    
    def get_name_by_ticker(self, ticker):
        df = self.to_dataframe()
        result = df[df["ticker"] == ticker]
        if not result.empty:
            return result.iloc[0]["name"]
        else:
            raise ValueError(f"No company found with ticker: {ticker}")
    
    def get_exchange_by_ticker(self, ticker):
        df = self.to_dataframe()
        result = df[df["ticker"] == ticker]
        if not result.empty:
            return result.iloc[0]["exchange"]
        else:
            raise ValueError(f"No company found with ticker: {ticker}")
    def get_tickers_by_exchange(self, exchange):
            """
            Get a list of tickers for companies listed on a specific exchange.
            
            Args:
                exchange (str): The name of the exchange to filter by.

            Returns:
                list: A list of tickers for companies on the specified exchange.
            """
            df = self.to_dataframe()
            result = df[df["exchange"].str.lower() == exchange.lower()]
            return result["ticker"].tolist()
    
    def get_all_exchanges(self):
            """
            Get a list of all unique exchanges in the dataset.

            Returns:
                list: A list of unique exchanges.
            """
            df = self.to_dataframe()
            return df["exchange"].dropna().unique().tolist()


    def get_companies_by_exchange(self, exchange):
        """
        Get a list of companies (CIK, name, and ticker) listed on a specific exchange.

        Args:
            exchange (str): The exchange name.

        Returns:
            list[dict]: A list of dictionaries containing CIK, name, and ticker for each company.
        """
        df = self.to_dataframe()
        result = df[df["exchange"] == exchange]
        if not result.empty:
            return result[["cik", "name", "ticker"]].to_dict(orient="records")
        else:
            raise ValueError(f"No companies found for exchange: {exchange}")



    def search_by_name(self, substring):
        df = self.to_dataframe()
        return df[df["name"].str.contains(substring, case=False)]
